test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java

```diff
@@ -0,0 +1,96 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2021 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package org.openj9.test.lworld;
+
+
+import jdk.internal.misc.Unsafe;
+import java.lang.reflect.Field;
+import static org.testng.Assert.*;
+import org.testng.annotations.Test;
+import org.testng.annotations.BeforeClass;
+import static org.openj9.test.lworld.ValueTypeUnsafeTestClasses.*;
+
+
+@Test(groups = { "level.sanity" })
+public class ValueTypeUnsafeTests {
+	static Unsafe myUnsafe = null;
+
+	@BeforeClass
+        static public void testSetUp() throws RuntimeException {
+                try {
+                        Field f = Unsafe.class.getDeclaredField("theUnsafe");
+                        f.setAccessible(true);
+                        myUnsafe = (Unsafe)f.get(null);
+                } catch (NoSuchFieldException | SecurityException | IllegalArgumentException | IllegalAccessException e) {
+                        throw new RuntimeException(e);
+		}
+	}
+
+
+	@Test
+        static public void testFlattenedFieldIsFlattened() throws Throwable {
+                ValueTypePoint2D p = new ValueTypePoint2D(new ValueTypeInt(0), new ValueTypeInt(0));
+                boolean isFlattened = myUnsafe.isFlattened(p.getClass().getDeclaredField("x"));
+                assertTrue(isFlattened);
+        }
+
+        @Test
+        static public void testFlattenedArrayIsFlattened() throws Throwable {
+                ValueTypePoint2D[] pointAry = {};
+                boolean isArrayFlattened = myUnsafe.isFlattenedArray(pointAry.getClass());
+                assertTrue(isArrayFlattened);
+        }
+
+        @Test
+        static public void testRegularArrayIsNotFlattened() throws Throwable {
+                Integer[] ary = {};
```

Comment:
Integer might be migrated to value type in the future. It is safer to use another non-value type class, e.g. `java.lang.Thread`.